Fix MJPEG camera decoding on Linux

shiena opened the pull request:
<!--
Please target the `master` branch in priority.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://contributing.godotengine.org/en/latest/engine/guidelines/code_style.html
-->
fixed #113357 
MJPEG streams from cameras have two issues:
1. V4L2 buffer size is allocated for the maximum frame size, which is larger than the actual JPEG data size. Passing the entire buffer to the decoder causes it to fail due to invalid trailing data.
2. Real-time streams may have partially corrupted frames. The standard JPEG decoder strictly rejects these as errors.

→ **Solution**:
- Use `v4l2_buffer.bytesused` to get the actual frame size.
- Add an internal function that decodes using libjpeg-turbo's lenient mode (tolerates partial corruption).

j20001970 commented on modules/jpg/image_loader_libjpeg_turbo.cpp:
```diff
@@ -68,6 +76,9 @@ Error jpeg_turbo_load_image_from_buffer(Image *p_image, const uint8_t *p_buffer,
 	if (tj3Decompress8(tj_instance, p_buffer, p_buffer_len, data.ptrw(), 0, tj_pixel_format) < 0) {
+		if (p_strict) {
+			print_verbose(vformat("libjpeg-turbo decompress error: %s", tj3GetErrorStr(tj_instance)));
+		}
```
From my testing that when decoding partially corrupted MJPEG frames, `tj3Decompress8` still returns -1 indicating a warning is encountered. [An issue from libjpeg-turbo](https://github.com/libjpeg-turbo/libjpeg-turbo/issues/157#issuecomment-312123754) also mentioned such behaviour but for `tjDecompress2`.

We can use `tjGetErrorCode` instead to determine if the error is fatal (`TJERR_FATAL`) or not (`TJERR_WARNING`).

j20001970 commented on modules/jpg/image_loader_libjpeg_turbo.cpp:
```diff
@@ -32,13 +32,21 @@
+		// Allow partial/corrupt JPEG decoding for streaming sources like cameras.
+		tj3Set(tj_instance, TJPARAM_STOPONWARNING, 0);
+	}
+
```
`TJPARAM_STOPONWARNING` is already 0 by default since libjpeg-turbo 2.1.90, maybe the additional `p_strict` parameter is not needed?

shiena commented:
It seems like the libjpeg-turbo upgrade in #113945 may have resolved the noise and frame drop issues. This PR might not be necessary anymore.

j20001970 reviewed (approved):
c36f7d4f13b6d2ff502138113738e91241180f6b works as expected with Logitech C270 webcam, no visual glitch is observed from decoded MJPEG frames.

shiena commented:
I tested with Steam Deck + Logitech C920 and there was no noise. The noise was occurring in a WSL2 + usbipd USB passthrough environment, so it appears the overhead was causing latency.

j20001970 commented:
Tested again with Logitech C270 webcam.

While MJPG decoding works, I'm curious about some decisions in 67af61291a5476a51854b2a6ae3f6c480a80f8e1:
- Why duplicating turbojpeg code in `JpegBufferDecoder`, when `jpeg_turbo_load_image_from_buffer` already do the (almost) same work?
- It seems that `Image::_load_jpg_from_buffer_lenient` ended up not used anywhere.

And after #119570 is merged, decoding partially corrupt MJPG frames with `Image::load_jpg_from_buffer` no longer result in errors, but it will spam warning about "extraneous bytes before marker" instead. I think this PR should be updated to better reflect the current situation in `master` branch.

shiena committed:
Fix MJPEG camera decoding with YUV plane output on Linux

Use v4l2_buffer.bytesused to skip trailing garbage from the
over-allocated capture buffer, decode straight to YUV planes for
GPU-side YUV->RGB conversion, and clear TJPARAM_STOPONWARNING to
tolerate partial frame corruption from streaming sources.

shiena commented:
@j20001970 
Yeah, you're right — those ended up as dead code, so I dropped _load_jpg_from_buffer_lenient, _jpg_mem_loader_func_ex, and the ImageMemLoadFuncEx typedef. core/io/image.* and modules/jpg/image_loader_libjpeg_turbo.cpp are back to matching master.

I did keep the direct turbojpeg calls in JpegBufferDecoder though, since tj3DecompressToYUVPlanes8 (YUV plane output for GPU-side YUV→RGB conversion) isn't exposed by jpeg_turbo_load_image_from_buffer, and reusing a single tjhandle across frames avoids per-frame setup cost on the streaming path. I added a short comment in buffer_decoder.cpp explaining this.
